### PR TITLE
Making default program deployment on RPC instead of QUIC

### PR DESCRIPTION
**Pull request by godmodegalactus**
#### Problem

QUIC is suboptimal to deploy program because it is hard to get unstaked connections.

#### Summary of Changes

We add option of deploy the program over RPC. By default it will use and RPC method unless we specify to deploy over QUIC using flag `--use-quic`.

### CODE REVIEW

**Comment by KirillLykov** [comment]
Would be nice to see comparison between old/new procedures in regard of speed and number of errors with stale blockhash (to verify that https://github.com/solana-labs/solana/issues/31203 has been resolved)

**Comment by KirillLykov** [comment]
CI fails with 

```sh
thread 'program::tests::test_cli_keypair_file' panicked at 'called `Result::unwrap()` on an `Err` value: "Data writes to account failed: Custom error: Max retries exceeded"', cli/src/program.rs:3153:56
```

**Review by t-nelson — CHANGES_REQUESTED** [review]
let's split the cli rework from the introduction of the new multi-tx send function

**Comment by godmodegalactus** [comment]
Don't make it stale. https://github.com/solana-labs/solana/pull/32638 with partial changes from this patch is under review.

File: cli/src/clap_app.rs [hunk 1]
```diff
@@ -84,19 +84,6 @@ pub fn get_clap_app<'ab, 'v>(name: &str, about: &'ab str, version: &'v str) -> A
...
-                .global(true)
-                .help("Use QUIC when sending transactions."),
-        )
-        .arg(
-            Arg::with_name("use_udp")
-                .long("use-udp")
-                .global(true)
-                .conflicts_with("use_quic")
-                .help("Use UDP when sending transactions."),
-        )
```
Comment (t-nelson) [inline cli/src/clap_app.rs:99]: we prefer to deprecate flags by hiding them, dropping their internal usage and printing a warning when they are used.  this prevents outright breaking consumers' workflow

File: client/src/send_and_confirm_transactions_in_parallel.rs [hunk 2]
```diff
@@ -456,6 +456,8 @@ pub async fn send_and_confirm_transactions_in_parallel<T: Signers + ?Sized>(
             break;
         }
 
+        tokio::time::sleep(Duration::from_millis(100)).await;
```
Comment (t-nelson) [inline client/src/send_and_confirm_transactions_in_parallel.rs:459]: unrelated?

File: cli/src/program.rs [hunk 3]
```diff
@@ -198,6 +219,19 @@ impl ProgramSubCommands for App<'_, '_> {
...
                                 .help("Use the designated program id even if the account already holds a large balance of SOL")
+                        )
+                        .arg(
+                            Arg::with_name("use_quic")
+                                .long("use-quic")
+                                .global(true)
+                                .help("Use QUIC when sending transactions."),
+                        )
+                        .arg(
+                            Arg::with_name("use_udp")
```
Comment (t-nelson) [inline cli/src/program.rs:230]: udp tpu is dead, there's no value in offering this option

File: cli/src/program.rs [hunk 4]
```diff
@@ -198,6 +219,19 @@ impl ProgramSubCommands for App<'_, '_> {
                                 .long("allow-excessive-deploy-account-balance")
                                 .takes_value(false)
                                 .help("Use the designated program id even if the account already holds a large balance of SOL")
+                        )
+                        .arg(
+                            Arg::with_name("use_quic")
+                                .long("use-quic")
+                                .global(true)
```
Comment (t-nelson) [inline cli/src/program.rs:226]: presumably the point of moving this arg declaration from the app to the subcommand was that it _not_ be global?

**Review by t-nelson — CHANGES_REQUESTED** [review]

File: client/src/send_and_confirm_transactions_in_parallel.rs (hunk 2 above)
Comment (godmodegalactus) [reply client/src/send_and_confirm_transactions_in_parallel.rs:459]: Yes unrelated. Will remove.

File: cli/src/program.rs (hunk 4 above)
Comment (godmodegalactus) [reply cli/src/program.rs:226]: It was just used in program deploy and write buffer. I will make it not global.

File: cli/src/clap_app.rs (hunk 1 above)
Comment (godmodegalactus) [reply cli/src/clap_app.rs:99]: After some tests with this PR i have following results : 

1. It is better to send the transactions to a private RPC in RPC only mode. 
2. RPC only mode performs much worse when deploying using public testnet rpc server.

I think this is because the testnet RPC is rate limited while it is still easy to get unstaked connections on testnet.

I think it will be better to add a flag `--rpc-only` so that deploying on mainnet with private rpc will be much faster. Meanwhile still use the old quic mode by default to deploy using testnet and localcluster.

Comment (t-nelson) [reply cli/src/clap_app.rs:99]: we need to optimize the defaults for the most commonly available configuration, not what works best. more sophisticated actors can be assumed competent to adjust their configuration as appropriate.

i don't see direct submission via quic-tpu in this analysis. where does it fit in reliability? intuitively ascending reliability would be something like `unstaked_quic_tpu` -> `public_rpc` -> `private_rpc` -> `staked_tpu`. which is inverse to availability. so i'd think defaulting to `public_rpc` which can be trivially upgraded to `private_rpc` with existing `-u RPC_URL` global arg, then `--use-quic` for those with staked quic seats to further improve reliability

**Comment by godmodegalactus** [comment]
Dont think the PR is needed anymore, the recent changes by Jon Cinque for https://github.com/solana-labs/solana/pull/32945 has made the performance much better.
